File: test/tracking/test_sort.cpp
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <nlohmann/json.hpp>
#include <opencv2/opencv.hpp>

#include "tracking/Sort.h"

int main() {
	nlohmann::json j;
	std::ifstream f(std::filesystem::path(CMAKE_SOURCE_DIR) / std::filesystem::path("test/tracking/detections_file.json"));
	f >> j;

	tracking::Sort<> tracker;

	std::int64_t timestamp_old = -1;
	for (auto frame : j["data"]) {
		if (frame["src"] != "s110_s_cam_8") continue;

		std::vector<tracking::Detection2D<tracking::BoundingBoxXYXY>> detections;
		for (auto const& detection : frame["detections"]) {
			detections.emplace_back(tracking::BoundingBoxXYXY(detection["left"], detection["top"], detection["right"], detection["bottom"]), detection["conf"], 1);
		}

		common::println("dt: ", static_cast<double>(frame["timestamp"].get<std::int64_t>() - timestamp_old) / 1000.);
		// first dt is undefined but also never used;
		auto matched = tracker.update(static_cast<double>(frame["timestamp"].get<std::int64_t>() - timestamp_old) / 1000., detections);

		auto img = cv::imread(std::filesystem::path(CMAKE_SOURCE_DIR) / std::filesystem::path("data/camera_simulator/s110_s_cam_8/s110_s_cam_8_images") / std::filesystem::path(to_string(frame["timestamp"]) + ".jpg"));

		for (auto const& detection : detections) cv::rectangle(img, cv::Point2d(detection.bbox().left(), detection.bbox().top()), cv::Point2d(detection.bbox().right(), detection.bbox().bottom()), cv::Scalar_<int>(0, 0, 0), 1);
		for (auto const& e : matched) {
			e.matched() ? cv::rectangle(img, cv::Point2d(e.bbox().left(), e.bbox().top()), cv::Point2d(e.bbox().right(), e.bbox().bottom()), cv::Scalar_<int>(0, 0, 255), 5)
			        : cv::rectangle(img, cv::Point2d(e.bbox().left(), e.bbox().top()), cv::Point2d(e.bbox().right(), e.bbox().bottom()), cv::Scalar_<int>(0, 0, 255), 1);

			cv::putText(img, std::to_string(e.id()), cv::Point2d(e.bbox().left(), e.bbox().top()), cv::FONT_HERSHEY_DUPLEX, 1.0, cv::Scalar_<int>(0, 0, 0), 1);
		}

		cv::imshow("img", img);
		cv::waitKey(0);

		timestamp_old = frame["timestamp"];
	}

	return 0;
}

File: test/tracking/test_sort.py
from scripts.sort import Sort
import json
import numpy as np
import cv2
import os
import random

from scripts.ocsort import OCSort


def main_ocsort():
    with open('detections_file.json') as file:
        j = json.loads(file.read())

    tracker = OCSort()
    for frames in j["data"]:
        if frames["src"] != "s110_w_cam_8": continue

        xs = np.array(
            [np.array([detection["left"], detection["top"], detection["right"], detection["bottom"]]) for detection in
             frames["detections"]])

        tracker.update(xs)



def main_sort():
    with open('detections_file.json') as file:
        j = json.loads(file.read())

    tracker = Sort(max_age=3, min_hits=3, iou_threshold=0.1)

    old_timestamp = -1
    for frames in j["data"]:
        if frames["src"] != "s110_w_cam_8": continue

        xs = np.array(
            [np.array([detection["left"], detection["top"], detection["right"], detection["bottom"]]) for detection in
             frames["detections"]])

        print("dt: ", (frames["timestamp"] - old_timestamp) / 1000 if old_timestamp > 0 else 0)
        tracks, unmatched_tracks, unmatched_dets = tracker.update(dets=xs, dt=(frames[
                                                                                   "timestamp"] - old_timestamp) / 1000 if old_timestamp > 0 else 0)

        img = cv2.imread("/home/lukas/src/minimal/data/camera_simulator/s110_w_cam_8/s110_w_cam_8_images/" + str(
            frames["timestamp"]) + ".jpg")

        for track in tracks:
            rng = np.random.default_rng(int(track[4]))
            random.seed(int(track[4]))
            cv2.rectangle(img, (int(track[0]), int(track[1])), (int(track[2]), int(track[3])),
                          (random.randint(0, 256), random.randint(0, 256), random.randint(0, 256)), 20)

        for track in unmatched_tracks:
            rng = np.random.default_rng(int(track[4]))
            random.seed(int(track[4]))
            cv2.rectangle(img, (int(track[0]), int(track[1])), (int(track[2]), int(track[3])),
                          (random.randint(0, 256), random.randint(0, 256), random.randint(0, 256)), 1)

        for track in unmatched_dets:
            cv2.rectangle(img, (int(track[0]), int(track[1])), (int(track[2]), int(track[3])),
                          (0, 0, 255), 20)

        for track in xs:
            cv2.rectangle(img, (int(track[0]), int(track[1])), (int(track[2]), int(track[3])),
                          (0, 0, 0), 1)

        cv2.imshow("img", img)
        cv2.waitKey(0)

        old_timestamp = frames["timestamp"]


if __name__ == '__main__':
    main_sort()
